File: frontend/src/app/core/services/product.service.ts
import { Injectable } from '@angular/core';
import { ApiService } from './api.service';
import { Observable } from 'rxjs';
import { Product } from '../models/product.model';

@Injectable({
    providedIn: 'root'
})
export class ProductService {
    private endpoint = '/products';
    constructor(private api: ApiService) { }

    getProducts(): Observable<Product[]> {
        return this.api.get<Product[]>(this.endpoint);
    }

    createProduct(product: Product): Observable<Product> {
        return this.api.post(this.endpoint, product);
    }

    updateProduct(id: string, product: Product): Observable<Product> {
        return this.api.put(`${this.endpoint}/${id}`, product);
    }

    deleteProduct(id: string): Observable<void> {
        return this.api.delete(`${this.endpoint}/${id}`);
    }

    getMyProducts(sellerId: string): Observable<Product[]> {
        return this.api.get<Product[]>(`${this.endpoint}/my-products`);
    }
}

File: microservices/product-service/src/main/java/com/buy01/product/service/ProductService.java
package com.buy01.product.service;

import com.buy01.product.dto.ProductRequest;
import com.buy01.product.model.Product;
import com.buy01.product.repository.ProductRepository;
import lombok.RequiredArgsConstructor;
import org.springframework.stereotype.Service;
import java.util.List;

@Service
@RequiredArgsConstructor
public class ProductService {

    private final ProductRepository repository;

    public Product createProduct(ProductRequest request, String sellerId) {
        Product product = Product.builder()
                .name(request.getName())
                .description(request.getDescription())
                .price(request.getPrice())
                .stockQuantity(request.getStockQuantity())
                .imageUrls(request.getImageUrls())
                .sellerId(sellerId) // Link to the user who created it
                .build();
        return repository.save(product);
    }

    public List<Product> getAllProducts() {
        return repository.findAll();
    }

    public List<Product> getMyProducts(String sellerId) {
        return repository.findBySellerId(sellerId);
    }

    public Product updateProduct(String id, ProductRequest request, String sellerId) {
        Product product = repository.findById(id)
                .orElseThrow(() -> new RuntimeException("Product not found"));

        if (!product.getSellerId().equals(sellerId)) {
            throw new RuntimeException("Access Denied: You are not the owner of this product");
        }

        product.setName(request.getName());
        product.setDescription(request.getDescription());
        product.setPrice(request.getPrice());
        product.setStockQuantity(request.getStockQuantity());
        product.setImageUrls(request.getImageUrls());

        return repository.save(product);
    }

    public void deleteProduct(String id, String sellerId) {
        Product product = repository.findById(id)
                .orElseThrow(() -> new RuntimeException("Product not found"));

        if (!product.getSellerId().equals(sellerId)) {
            throw new RuntimeException("Access Denied: You are not the owner of this product");
        }

        repository.delete(product);
    }
}
